chore: configure neostandard + prettier

bajtos opened the pull request:
I copied eslint and prettier configuration from https://github.com/CheckerNetwork/assert-ok-response

I split the changes into two commits to make this pull request easier to review. The second commit contains linting fixes automatically generated by eslint and prettier.

Last but not least, I also reworked npm scripts for linting to be consistent with other repositories.

- **chore: configure neostandard + prettier**
- **chore: `npm run lint:fix`**

Links:
- https://github.com/CheckerNetwork/roadmap/issues/166

juliangruber commented on .prettierrc.yaml:
```diff
@@ -1,2 +1,5 @@
-printWidth: 80
+semi: false
+singleQuote: true
+printWidth: 100
```
The PR this was copied from didn't get reviewed, therefore I believe here is the place where we need to discuss standards that work for the team. @bajtos can you please manage that process?

Things I would like to change:
- printWidth: **80** (tradition, enforces low nesting)
- trailingComma: **none** (I personally really don't like trailing commas)

As a compromise as always, I like using tools without changing their configuration. So, if we can just use prettier defaults, I'll +1 it (even though it uses trailing commas). This will help keep our code base opinion free, and follow community standards.

juliangruber replied:
I realize I had already approved https://github.com/CheckerNetwork/spark-checker/pull/127. This was for one repo though. If we copy this exact config to every repo, I would like to have a discussion about it first please.

bajtos replied:
SGTM, let's discuss the desired style here 👍🏻 

**trailing comma**

Some prior discussions happened in https://github.com/CheckerNetwork/roadmap/issues/166

> > [bajtos] Most importantly, Standard does not allow trailing commas in arrays & object literals. I am frustrated by this several times a day. https://github.com/standard/standard/issues/240
>
> [juliangruber] When it comes to form, I dislike them. When it comes to function, they are keeping git diffs clean, so I'm in favor.

FWIW, `trailingComma: all` is the current default behaviour of Prettier. (The default value changed in Prettier v3.0.0.) See https://prettier.io/docs/options#trailing-commas

**text width**

See https://prettier.io/docs/options#print-width

- The default is 80 characters, as you suggested, @juliangruber 
- Prettier recommends against using more than 80 characters (see the doc page I linked to)

juliangruber replied:
Cool, I'm happy with default trailing comma and default max width

pyropy replied:
I really like trailing commas👉🏻👈🏻 

Other precerences: single quotes, shorther print width (80 seems okay) to avoid nesting.

NikolasHaimerl replied:
I am with Julian on this one. Using the default settings usually gives you the experience the creator of a tooling had in mind and that usually works well. It's also the easiest to keep consistent across repos.

bajtos commented on .prettierrc.yaml:
```diff
@@ -1,2 +1,5 @@
-printWidth: 80
+semi: false
```
Should we keep using no semicolons or should we switch to Prettier defaults and start using semicolons again?

https://prettier.io/docs/options#semicolons

I slightly prefer to use semicolons for consistency with Rust, where a statement with a semicolon has a different meaning than a statement without one.

I believe the default TypeScript formatter uses semicolons too. But they also use double quotes, IIRC, while we are using single quotes, so this may not be the best example.

juliangruber replied:
My preference is without semicolons

pyropy replied:
I would vote in favor of no semicolons. Personally I consider code with no semicolons more readable.

bajtos commented on .prettierrc.yaml:
```diff
@@ -1,2 +1,5 @@
-printWidth: 80
+semi: false
+singleQuote: true
```
Do we prefer single quotes or should we switch to Prettier defaults - double quotes?

https://prettier.io/docs/options#quotes

bajtos replied:
My preference: use single quotes

bajtos commented on .prettierrc.yaml:
```diff
@@ -1,2 +1,5 @@
+semi: false
+singleQuote: true
+printWidth: 100
+trailingComma: all
```
`trailingComma: all` is Prettier's default, we are discussing this in https://github.com/CheckerNetwork/node/pull/710/files#r2024437202

bajtos commented on .prettierrc.yaml:
```diff
@@ -1,2 +1,5 @@
+singleQuote: true
+printWidth: 100
+trailingComma: all
 proseWrap: always
```
How do we want to wrap Markdown text?

https://prettier.io/docs/options#prose-wrap

I believe we are already using `proseWrap: always` in all our repositories.

By default, Prettier does not apply any line wrapping rules for Markdown text.

bajtos replied:
My preference: keep `proseWrap: always`.

Long lines are difficult to review on GitHub, as a comment for one line applies to the entire paragraph on that single line.

juliangruber replied:
I find diffs very noisy with wrapping, and prefer without wrapping

bajtos commented:
I updated the Prettier config based on our discussions above, removed settings where we use Prettier defaults and kept only settings where we differ.

The new Prettier config:

```yaml
semi: false
singleQuote: true
```

After reviewing the changes, I added two more files to `.prettierignore`:

- `/lib/abi.json` - this is an auto-generated definition of a smart-contract ABI we copy from elsewhere, IIRC.
- `/tsconfig.json` - let's keep the list of included/excluded files formatted as one entry per line to reduce noisy diffs in the future.

I consider this PR ready for final review & merging.

juliangruber commented:
> /tsconfig.json - let's keep the list of included/excluded files formatted as one entry per line to reduce noisy diffs in the future.

Can we instead //ignore that one line, and lint the rest? I don't like this special case. You can make the same argument for other files with arrays. Instead, should we tweak the config to always list array elements on multiple lines?

bajtos commented:
> > /tsconfig.json - let's keep the list of included/excluded files formatted as one entry per line to reduce noisy diffs in the future.
> 
> Can we instead //ignore that one line, and lint the rest? I don't like this special case. You can make the same argument for other files with arrays. Instead, should we tweak the config to always list array elements on multiple lines?

To my best knowledge:
- You cannot add ignore directives to JSON files, because JSON does not support comments.
- I could not find any config option to tell Prettier to use one line per array item in JSON files

I see two conflicting requirements here:
- We want to avoid noisy diffs
- We don't want special cases

IIUC, avoiding special cases is more important to you than avoiding noisy diffs. Is that correct?

I don't have a strong opinion myself, I am going to remove `tsconfig.json` from `.prettierignore`

bajtos commented:
> I am going to remove `tsconfig.json` from `.prettierignore`


See [9c67ebb](https://github.com/CheckerNetwork/node/pull/710/commits/9c67ebbc8f0fe4981262bc39072edb9d6904192b)

juliangruber commented:
Have you looked into the other option I suggested, @bajtos? To configure the linter to always multi-line arrays, so that we avoid all of these cases?

bajtos commented:
> Have you looked into the other option I suggested, @bajtos? To configure the linter to always multi-line arrays, so that we avoid all of these cases?

Yes, I did. Quoting from https://github.com/CheckerNetwork/node/pull/710#issuecomment-2777709367, with added emphasis:

> To my best knowledge:
> 
> * You cannot add ignore directives to JSON files, because JSON does not support comments.
> * **I could not find any config option to tell Prettier to use one line per array item in JSON files**

You can find the list of all Prettier options here:

https://prettier.io/docs/options

juliangruber commented:
Oh sorry I missed that. 😅 

There's https://www.npmjs.com/package/prettier-plugin-multiline-arrays as an option, if you agree I'm happy to add it to this PR.

Btw, I think we should soon think about bundling our linting logic into a module, so we can share the configuration among all our repos, and get dependabot updates for them. I'm also happy to work on that

bajtos commented:
> Oh sorry I missed that. 😅
> 
> There's https://www.npmjs.com/package/prettier-plugin-multiline-arrays as an option, if you agree I'm happy to add it to this PR.

Nice find!

If we are going to add prettier plugins, how do you feel about adding the following plugins too? (Discovered via https://github.com/electrovir/prettier-plugin-multiline-arrays?tab=readme-ov-file#compatibility):

- https://www.npmjs.com/package/prettier-plugin-packagejson
- https://www.npmjs.com/package/prettier-plugin-organize-imports
- prettier-plugin-jsdoc

> Btw, I think we should soon think about bundling our linting logic into a module, so we can share the configuration among all our repos, and get dependabot updates for them. I'm also happy to work on that

Great idea! I propose the following next steps:

1. Finish the discussion about the desired Neostandard & Prettier config in this pull request.
2. Land this PR.
3. Create a shared configuration
4. Update this repo to use the new shared configuration
5. Continue updating other repositories.

I propose creating two repositories for the shared config:
- CheckerNetwork/eslint-config
- CheckerNetwork/prettier-config

Thoughts?

juliangruber commented:
> If we are going to add prettier plugins, how do you feel about adding the following plugins too? (Discovered via https://github.com/electrovir/prettier-plugin-multiline-arrays?tab=readme-ov-file#compatibility):

> https://www.npmjs.com/package/prettier-plugin-packagejson

+1

> https://www.npmjs.com/package/prettier-plugin-organize-imports

+100

> prettier-plugin-jsdoc

https://www.npmjs.com/package/prettier-plugin-jsdoc

+100



> > Btw, I think we should soon think about bundling our linting logic into a module, so we can share the configuration among all our repos, and get dependabot updates for them. I'm also happy to work on that
> 
> Great idea! I propose the following next steps:
> 
> 1. Finish the discussion about the desired Neostandard & Prettier config in this pull request.
> 2. Land this PR.
> 3. Create a shared configuration
> 4. Update this repo to use the new shared configuration
> 5. Continue updating other repositories.

💯 

> I propose creating two repositories for the shared config:
> 
> * CheckerNetwork/eslint-config
> * CheckerNetwork/prettier-config
> 
> Thoughts?

I had assumed we can create a binary, so it's just

```json
{
  "dependencies": {
    "@space-meridian/lint": "0.0.0"
  },
  "scripts": {
    "test:lint": "space-lint"
  }
}
```

This way, the project doesn't need to bother and maintained tool versions (prettier & eslint). I do wonder about whether we need extensibility of the configuration, and therefore the approach you suggested works better. Wdyt?

bajtos commented:
> I had assumed we can create a binary

Unfortunately, that does not work with IDEs like VSCode and Zed. There was a long discussion between Neostandard and Eslint maintainers on this topic, see e.g. https://github.com/neostandard/neostandard/pull/2, https://github.com/neostandard/neostandard/issues/33 and other linked issues/PRs.

bajtos commented:
> https://www.npmjs.com/package/prettier-plugin-organize-imports
>
> +100

I am not able to get this working. Let's leave it up for future iterations.

bajtos commented:
I added a bunch of commits to configure additional Prettier plugins:
- prettier-plugin-multiline-arrays
- prettier-plugin-packagejson
- prettier-plugin-jsdoc

For each plugin, there are two commits - one with the configuration and another with the formatting changes. I hope it will make it easier to review this pull request.

juliangruber commented:
> I added a bunch of commits to configure additional Prettier plugins:
> 
> * prettier-plugin-multiline-arrays
> * prettier-plugin-packagejson
> * prettier-plugin-jsdoc
> 
> For each plugin, there are two commits - one with the configuration and another with the formatting changes. I hope it will make it easier to review this pull request.

FYI, for me, I usually don't review individual commits, as you can't add comments. With the full PR view some things are a bit harder to see, but I prefer that approach for any PR that isn't thousands of lines.

juliangruber commented on commands/checker.js:
```diff
@@ -13,14 +13,9 @@ import { ethers, formatEther } from 'ethers'
-const runtimeNames = [
-  'zinnia'
-]
+const runtimeNames = ['zinnia']
```
Interesting, didn't we want things like this to be multi line arrays?

juliangruber replied:
Do we need to set the threshold to 0?

bajtos replied:
I tried to set the threshold to `0`, but that made other lines look ugly. I will add a commit to show the result.

bajtos replied:
Reverted in https://github.com/CheckerNetwork/node/pull/710/commits/1c1ff2b0e751545587c8cfbb14a11447d112f16d

juliangruber commented on lib/metrics.js:
```diff
@@ -41,65 +42,71 @@ export class Metrics {
+    for (const [
+      ,
+      metrics,
+    ] of this.subnetMetrics) {
```
This looks quite bad. Restructuring with multiline arrays is pretty ugly :|

juliangruber commented on lib/metrics.js:
```diff
@@ -41,65 +42,71 @@ export class Metrics {
+        ([
+          key,
+          value,
+        ]) => mergedMetrics[key] !== value,
```
Same here wrt destructuring and multi line arrays

juliangruber replied:
Maybe this plugin won't work so well after all, and maybe that's why it's not included in prettier core

juliangruber commented on scripts/version.js:
```diff
@@ -7,4 +7,7 @@ import * as paths from '../lib/paths.js'
+await execa('git', [
+  'add',
+  'package.json',
+])
```
Reading this PR more and more, I think I would like to remove the multi line array plugin. The desire to keep diffs clean is imposing a too inconvenient style

bajtos commented on commands/checker.js:
```diff
@@ -67,42 +78,55 @@ export const checker = async ({ json, recreateCheckerIdOnError, experimental })
-        rewardsScheduledForAddress:
-          formatEther(metrics.rewardsScheduledForAddress)
-      }, null, 2))
+      console.log(
+        JSON.stringify(
+          {
+            totalJobsCompleted: metrics.totalJobsCompleted,
+            rewardsScheduledForAddress: formatEther(
+              metrics.rewardsScheduledForAddress,
+            ),
+          },
+          null,
+          2,
+        ),
+      )
```
This would be better formatted if we used `%o` instead of JSON.stringify, I think.

```js
console.log('%o', {
 totalJobsCompleted: metrics.totalJobsCompleted,
 rewardsScheduledForAddress: formatEther(
 metrics.rewardsScheduledForAddress,
})
```

I feel such a change is out of the scope of this PR. WDYT?

pyropy reviewed (approved):
Trailing commas woo-hoo! 🎉 

Overall this looks very good to me. I don't have any strong opinions on how the code is formatted now but I'm glad we have settled on styling configuration. 👍🏻

juliangruber reviewed (approved):
![love it](https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExcGoxZjExeXk1bnI2OG9wd2J1aGYxNGt1d3RnYTF2eXg2NW41MmgycSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/xT0xeIbYVQcBFDSdVu/giphy.gif)
